src/app.js

```diff
@@ -211,40 +213,107 @@ export default function (app) {
   }
 
-  function patch(parent, element, oldNode, node) {
+  function batchRemove (parent, element, node) {
+      batch.push(parent.removeChild.bind(parent, element))
+      if (node.data && node.data.onRemove) {
+        defer(node.data.onRemove, element)
+      }
+  }
+
+  function patch (parent, element, oldNode, newNode) {
     if (oldNode == null) {
-      element = parent.appendChild(createElementFrom(node))
+      var n = createElementFrom(newNode)
+      parent.insertBefore(n, element)
+      element = n
+
+    } else if (
+      newNode.tag !== oldNode.tag ||
+      typeof newNode !== typeof oldNode ||
+      typeof newNode === "string" && newNode !== oldNode
+    ) {
+      var n = createElementFrom(newNode)
+      parent.replaceChild(n, element)
+      element = n
+
+    } else if (oldNode.tag){
+      updateElementData(element, newNode.data, oldNode.data)
+      patchChildren(element, oldNode, newNode)
+    }
 
-    } else if (node == null) {
-      batch.push(function () {
-        if (oldNode.data && oldNode.data.onRemove) {
-          oldNode.data.onRemove(element)
-        }
-        parent.removeChild(element)
-      })
-    } else if (node.tag && node.tag === oldNode.tag) {
-      updateElementData(element, node.data, oldNode.data)
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var oldChildren = []
+    var oldKeys = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      oldNode.children[i].data &&
+      oldNode.children[i].data.key &&
+      (oldKeys[oldNode.children[i].data.key] = true);
+      oldChildren[i] = [element.childNodes[i], oldNode.children[i]]
+    }
 
-      var len = node.children.length
-      var oldLen = oldNode.children.length
+    var isKeyed = function (node) {
+      return node.data && node.data.key && oldKeys[node.data.key]
+    }
 
-      for (var i = 0; i < len || i < oldLen; i++) {
-        patch(element, element.childNodes[i], oldNode.children[i], node.children[i])
-      }
+    var newIndex = 0
+    var oldIndex = 0
+    var cache = {}
+
+    while (
+      newIndex < newNode.children.length ||
+      oldIndex < oldChildren.length
+    ) {
+      var newChild = newNode.children[newIndex]
+      var oldChild = oldChildren[oldIndex][1]
+
+      if (
+        newChild != null &&
+        isKeyed(newChild) &&
+        cache[newChild.data.key]
+      ) {
+        var i = cache[newChild.data.key]
+        element.insertBefore(i[0], oldChildren[oldIndex][0])
+        patch(element, i[0], i[1], newChild)
+        delete cache[newChild.data.key]
+        newIndex++
+
+      } else if (oldChild == null || (isKeyed(oldChild) && !isKeyed(newChild))) {
+        patch(element, oldChildren[oldIndex][0], null, newChild)
+        newIndex++
+
+      } else if (newChild == null || (!isKeyed(oldChild) && isKeyed(newChild)) ) {
```

Comment:
I tried that, but that change actually *added* 11 bytes. Either way I'll leave the change in the upcoming commit, so that you can see the result and perhaps help me find a shorter way to write it.